### PR TITLE
docs: add docs to functions and setup auto-generation of docs

### LINKED ISSUE #56: 📝 Documentation: Split API docs out from README.md
### Bug Report Checklist

- [X] I have pulled the latest `main` branch of the repository.
- [X] I have [searched for related issues](https://github.com/JoshuaKGoldberg/ts-api-utils/issues?q=is%3Aissue) and found none that matched my issue.

### Overview

From @RebeccaStevens' comment in https://github.com/JoshuaKGoldberg/ts-api-utils/issues/52#issuecomment-1427642438:

> Due to the share number of functions exported by this library, I think it would be better to put all these docs in an `API.md` file rather than cluttering up the readme. The readme can then link to that file. Each of the categories within the API docs could even be split up into sub-files to prevent the API file from becoming huge.

Agreed on splitting up. My default instinct is to create a `docs/` directory, like how [TypeStat's docs are](https://github.com/JoshuaKGoldberg/TypeStat/tree/7121613bf60381d588af7d2b8833859444fd50bc/docs). What do you think?
 

### Additional Info

Maybe this is a precursor to creating a live site like https://tsup.egoist.dev? I wouldn't mind have a `ts-api-tools.joshuakgoldberg.dev`... 😄

### PR DESCRIPTION
**Pull request by RebeccaStevens**
## PR Checklist

- [x] Addresses an existing open issue:
  - fixes #56
  - closes #42 (testing should be necessary as docs are auto-generated)
- [x] That issue was marked as [`status: accepting prs`](https://github.com/JoshuaKGoldberg/ts-api-utils/issues?q=is%3Aopen+is%3Aissue+label%3A%22status%3A+accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/JoshuaKGoldberg/ts-api-utils/blob/main/.github/CONTRIBUTING.md) were taken

## Overview

- Added a bunch of JSDoc comments to the functions (didn't touch the "type" code yet as waiting for #46).
- Configured [typedoc](https://typedoc.org/) to generate docs from those JSDoc comments.
- Set up a workflow to automatically upload the generated docs to GitHub Pages after each release.

### CODE REVIEW

File: typedoc.json
Comment (JoshuaKGoldberg) [inline typedoc.json:33]: [Comment] I have a few miscellaneous feature requests that are definitely not blocking for this PR and probably shouldn't go in it even if they're immediately actionable:

* Integrating [Konamimojisplosion](https://github.com/JoshuaKGoldberg/Konamimojisplosion) 😁 
* Adding a <kbd>Ctrl</kbd>/<kbd>Cmd</kbd>+<kbd>K</kbd> shortcut for search the way Docusaurus and a lot of other sites / site generators have
* Adding lint rules around tags: avoiding blank/redundant ones and including useful on exported things

...just posting here so I don't forget them.

File: typedoc.json
Comment (JoshuaKGoldberg) [inline typedoc.json:32]: [Request] Although I agree the full README.md's contents probably aren't what we want for the deployed website, it does feel a little confusing to have ... no introduction at all:

Screenshot of the typedoc-generated index page with links and no sentence descriptions

Can we have the section of the README.md that mentions usage and what this package is? I don't see a way to do this other than either [calling TypeDoc programmatically](https://typedoc.org/guides/installation/#node-module) _(eek!)_ or having a dedicated `.md` file. The dedicated file seems best to me?

File: typedoc.json
Comment (JoshuaKGoldberg) [inline typedoc.json:23]: [Question] Is there a way to hide the _Modules_ section from the nav bar? It's only ever going to be one in this package.

Screenshot of the Modules nav section with the one member "ts-api-utils - v0.0.22"

[Request] If there is, can we do it here? If not, we should file a followup issue to track requesting an option for it.

File: typedoc.json
Comment (JoshuaKGoldberg) [inline typedoc.json:17]: [Nit] This is the default, no? https://typedoc.org/guides/options/#entrypointstrategy

```suggestion
```

Same with `cleanOutputDir` I think?

File: .github/workflows/release-docs.yml
Comment (JoshuaKGoldberg) [inline .github/workflows/release-docs.yml:73]: [Praise] This is a great `.yml` file, thank you 😄. I've been meaning to look into GH page deployments for https://github.com/OpenContributionsProject/opencontributionssite/issues/29...

cc @blackgirlbytes, good timing here!

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
👏 @RebeccaStevens this is fantastic, thanks for sending it in! The site works great and I like the deployment strategy a lot. Also strong +1 to a lot of the customizations in `typedoc.json`

Mostly requesting changes around removing redundant tags & filling in some more site info.

File: README.md
Comment (JoshuaKGoldberg) [inline README.md:20]: [Request] 🤔 I'm not sure how to get this to work. `docs/` is gitignored so this is going to look odd on github.com. Is the right thing linking to the deployed GitHub site?

File: package.json
Comment (JoshuaKGoldberg) [inline package.json:32]: [Request] These two new commands should be mentioned in `.github/DEVELOPMENT.md`. `docs` is pretty essential and there's an existing list of `lint:*` commands.

File: .github/workflows/release-docs.yml
Comment (RebeccaStevens) [reply .github/workflows/release-docs.yml:53]: This is used for keeping multiple versions of the docs. The point of it is to download the current docs before creating the new ones. This essentially merges the two together.

You can test this out by running `pnpm run docs`, then changing the version in `package.json` and running `pnpm run docs` again.

File: src/comments.ts
Comment (RebeccaStevens) [reply src/comments.ts:11]: I think there is a plugin that lets us categories by folder structure. Haven't tested it though.

File: knip.jsonc
Comment (RebeccaStevens) [reply knip.jsonc:8]: Will do. A new typedoc plugin for knip will be needed to handle these.

File: src/comments.ts
Comment (RebeccaStevens) [reply src/comments.ts:49]: A lot of these were just auto-generated when I created the JSDoc and I never removed the non necessary parts.

File: typedoc.json
Comment (RebeccaStevens) [reply typedoc.json:32]: Do you want write that file?

File: typedoc.json
Comment (RebeccaStevens) [reply typedoc.json:23]: I don't think there is. At least now without a plugin.

File: typedoc.json
Comment (RebeccaStevens) [reply typedoc.json:17]: We can remove them if you like

File: README.md
Comment (RebeccaStevens) [reply README.md:20]: Ah right. An oversight by me. I'll fix that up.

File: src/comments.ts
Comment (RebeccaStevens) [reply src/comments.ts:13]: No they aren't required. I think it's a good idea to still have them though.
The `@category` should be removed.
We can use an `@internal` if we want to be explicit that the function is only internally used.

File: .github/workflows/release-docs.yml
Comment (JoshuaKGoldberg) [reply .github/workflows/release-docs.yml:58]: Oh! That is snazzy. I like it 😄

File: typedoc.json
Comment (JoshuaKGoldberg) [reply typedoc.json]: Can do! For now I think it's fine to just copy & paste a sentence or three from the README.md. I can take this on as a followup.

File: src/types/getters.ts
Comment (RebeccaStevens) [inline src/types/getters.ts:59]: @JoshuaKGoldberg would you be able to write a quick doc for this function?
I'm not sure how to word the "WellKnownSymbol" thing?

File: typedoc.json
Comment (RebeccaStevens) [reply typedoc.json:41]: I don't think this does any thing for us. I believe those settings only effect classes (maybe interfaces?).

File: typedoc.json
Comment (RebeccaStevens) [reply typedoc.json:3]: I made this change.

File: .github/DEVELOPMENT.md
Comment (JoshuaKGoldberg) [inline .github/DEVELOPMENT.md:33]: [Nitpick]

```suggestion
- `pnpm lint:docs` ([typedocs](https://typedoc.org)): Checks that all the code is documented
```

File: .github/DEVELOPMENT.md
Comment (JoshuaKGoldberg) [inline .github/DEVELOPMENT.md:35]: Ah, I forgot about this one 😄

File: .github/DEVELOPMENT.md
Comment (JoshuaKGoldberg) [inline .github/DEVELOPMENT.md:70]: [Nitpick] Grammar, and a bit more info.

```suggestion
You can generate a [TypeDoc](https://typedoc.org) documentation site with `pnpm docs`.
```

File: .github/DEVELOPMENT.md
Comment (JoshuaKGoldberg) [inline .github/DEVELOPMENT.md:71]: [Question] I haven't been using one - is there a real benefit to?

[Request] If not, how about:

```suggestion
To browse the docs locally, open `docs/dev/index.html` in a web browser.
```

File: .github/actions/git-config/action.yml
Comment (JoshuaKGoldberg) [inline .github/actions/git-config/action.yml:2]: [Typo]

```suggestion
description: Configure the git user
```

File: src/types/getters.ts
Comment (JoshuaKGoldberg) [reply src/types/getters.ts:59]: How about _`* Retrieves a type symbol corresponding to a well-known string name.`_ for now?

File: typedoc.json
Comment (JoshuaKGoldberg) [inline typedoc.json:52]: [Question] Hmm, I see `docs/` as a more standard folder locally. Most projects I've seen set up something like this use `docs/`. `gh-pages` is usually the branch name. Is there a reason to prefer this?

File: typedoc.json
Comment (JoshuaKGoldberg) [inline typedoc.json:15]: [Suggestion] It might make sense to move the file into `src/`? The root path is normally for config files. But I don't feel strongly here, certainly not enough to block.

**Review by JoshuaKGoldberg — APPROVED** [review]
💯 awesome, thanks @RebeccaStevens! This is going to be great. Ready to ship whenever you want - only some suggestions on my end, nothing blocking that's bigger than a small nitpick/typo.

**Comment by JoshuaKGoldberg** [comment]
LMK if the branch protections are still messed up, I thought I fixed them but you never know.

File: .github/DEVELOPMENT.md
Comment (RebeccaStevens) [reply .github/DEVELOPMENT.md:71]: You can view the pure html of `docs/*/index.html` files with the `file://` protocol, but some features that rely on JavaScript aren't available with this protocol (due to CORS policy). One being the version selector.

`docs/index.html` itself isn't viewable with the `file://` protocol.

File: typedoc.json
Comment (RebeccaStevens) [reply typedoc.json:52]: I made this change simply so that we can write our own source docs files in the `docs` folder.

We're also using the newer way of using GitHub Pages that doesn't require us to push any built doc files to the repo. So ultimately the folder we build the docs in to doesn't matter (so long as nothing else is using it - aka it's empty) - the folder is never committed.

File: typedoc.json
Comment (RebeccaStevens) [reply typedoc.json:15]: I don't like the idea of putting it in `src/` as it's not a source file for our library itself. Maybe putting it in `docs/` instead.

File: src/types/getters.ts
Comment (RebeccaStevens) [inline src/types/getters.ts:102]: @JoshuaKGoldberg Do you want to doc this one to?

**Comment by RebeccaStevens** [comment]
> LMK if the branch protections are still messed up, I thought I fixed them but you never know.

I don't seem to have the option to force a merge, only enable auto merge / merge when all required checks pass.

image

**Comment by RebeccaStevens** [comment]
I want to resolve #68 and #83 first before merging this. That should result in all the checks passing.

File: src/nodes/typeGuards/compound.ts
Comment (JoshuaKGoldberg) [inline src/nodes/typeGuards/compound.ts:36]: Nit: I've always preferred _"whether"_ instead of _"true/false if"_ for boolean descriptions. Fewer words to work with. Not a blocker IMO.

```suggestion
 * @returns Whether the given node appears to be a {@link ConstAssertionExpression}.
```

_(here and elsewhere)_

File: src/nodes/typeGuards/compound.ts
Comment (JoshuaKGoldberg) [inline src/nodes/typeGuards/compound.ts:35]: [Nitpick] I don't think this description adds anything.

```suggestion
```

_(here and elsewhere)_

File: .github/DEVELOPMENT.md
Comment (JoshuaKGoldberg) [reply .github/DEVELOPMENT.md:71]: Ah, got it. I'd suggest adding in a `package.json` command like `"docs:serve": "http-serve docs/generated` then. But if you have a strong preference for this then I'm down to try it out 😄

File: typedoc.json
Comment (JoshuaKGoldberg) [reply typedoc.json:52]: Gotcha. I'd vote for `docs/generated/` as the place to put them, but not a blocker from me.

**Review by JoshuaKGoldberg — APPROVED** [review]
❤️‍🔥 awesome! Really excited to get this in.

File: .github/DEVELOPMENT.md
Comment (RebeccaStevens) [reply .github/DEVELOPMENT.md:71]: We can add that command if you like.
I was just using "Live Preview" because I already had that extension installed. You just have to right click on an html file and select "Show Preview".

File: src/types/getters.ts
Comment (RebeccaStevens) [reply src/types/getters.ts:104]: Actually, this one is internal. We don't *need* to add doc comment.

File: src/types/utilities.ts
Comment (JoshuaKGoldberg) [inline src/types/utilities.ts:167]: [Nitpick] Just noticed this 😄 

```suggestion
 * Returns true for `Object.defineProperty(o, 'prop', {value, writable: false})` and `Object.defineProperty(o, 'prop', {get: () => 1})`
```

File: .github/DEVELOPMENT.md
Comment (JoshuaKGoldberg) [inline .github/DEVELOPMENT.md:70]: On my Mac, this just opens the GitHub site. 

```suggestion
You can generate a [TypeDoc](https://typedoc.org) documentation site with `pnpm run docs`.
```

File: .github/DEVELOPMENT.md
Comment (JoshuaKGoldberg) [inline .github/DEVELOPMENT.md:71]: [Typo]

```suggestion
Once you've generated the docs, you can view them by first running `pnpm docs:serve`, then browsing to <http://127.0.0.1:8080> (or similar - check the console output of the last command).
```

File: .github/DEVELOPMENT.md
Comment (JoshuaKGoldberg) [inline .github/DEVELOPMENT.md:72]: [Suggestion] I've found this to be helpful:

```suggestion

> Tip: Run `pnpm run docs --watch` to keep the generated docs up-to-date as you save files.

```

File: .github/workflows/release-docs.yml
Comment (JoshuaKGoldberg) [inline .github/workflows/release-docs.yml:64]: Out of curiosity, why not the latest stable?

**Review by JoshuaKGoldberg — APPROVED** [review]
🔥 This is fantastic, thanks so much @RebeccaStevens! I learned a bunch of things from this PR 😄.

Ready to merge whenever you're happy. 🚀

**Comment by RebeccaStevens** [comment]
I can't guarantee that the workflow is completely correct. We might have to debug it once this is merged.

File: .github/workflows/release-docs.yml
Comment (RebeccaStevens) [reply .github/workflows/release-docs.yml:64]: No real reason, either would work.

Comment (JoshuaKGoldberg) [reply .github/workflows/release-docs.yml:64]: I'd vote for the stable then

**Comment by RebeccaStevens** [comment]
I can't merge this due to unresolved conversations. I'll resolve them then unresolve them so I can merge this.

File: typedoc.json
Comment (JoshuaKGoldberg) [reply typedoc.json]: #145 -> #149

**Pull request merged**
